### PR TITLE
Replacing usage of RocksDB timestamps 

### PR DESCRIPTION
**Pull request by cloudnoize**
In this PR, we revert the usage of RocksDB timestamps due to several bugs in their implementation.
Therefore, we no longer need to upgrade the RocksDB version, we revert to version 6.8.1.

Instead of RocksDB timestamps we save the version as the postfix of the value.

The challenge is to perform delete last reachable for recovery.
We introduce the following recovery sequence:

- Upon any BFT sequence number change, take a RocksDB snapshot and save its sequence number along with the BFT sequence number.
- In case of a crash, open the blockchain in read-only mode in order to disable possible compaction that will corrupt the snapshot that was taken in the previous step.
- open a recovery DB, a new RocksDBdb used to store the recovery data.
- for each block, we want to delete, take the updates needed in order to delete it using the RocksDB snapshot and save them in the recovery DB.
- Close read-only blockchain and recovery DB.
- open blockchain in write mode, in case replica state sync decides that deletion of last blocks is needed, the delete last block will open the recovery DB and read the updates required to delete the last block.
- the recovery DB is immutable, and only after successful completion of the deletion it will be erased. This fact is essential since it makes the flow resilient to crashes, as repeating any step will not have any side effects as long as we don't allow to update/change the recovery DB recovery information after it was first populated.

### CODE REVIEW

File: kvbc/include/v4blockchain/detail/column_families.h [hunk 1]
```diff
@@ -17,11 +17,6 @@
 
 namespace concord::kvbc::v4blockchain::detail {
 
-// Blockchain
-inline const auto BLOCKS_CF = std::string{"v4_blocks"};
-inline const auto ST_CHAIN_CF = std::string{"v4_st_chain"};
-inline const auto LATEST_KEYS_CF = std::string{"v4_latest_keys"};
-inline const auto IMMUTABLE_KEYS_CF = std::string{"v4_immutable_keys"};
-inline const auto CATEGORIES_CF = std::string{"v4_categories"};
+#include "storage/db_column_families.h"
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/detail/column_families.h:20]: As the timestamp changes are not used so the key comparator setting may not be needed. So it will be good to keep the column families here.

**Review by arc-vmware — COMMENTED** [review]
Thanks @cloudnoize , very nice changes. Completed till latest_keys.h file. I will continue again from there in an hour.

**Review by arc-vmware — APPROVED** [review]
@cloudnoize I had a first round of reading the code. The code is great. I have more questions and comments which I will continue to provide this week. But as this is important we should go ahead and commit the same.

File: kvbc/include/v4blockchain/detail/column_families.h (hunk 1 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/detail/column_families.h:20]: yes, we have debt with the compaction filter that we need to set, so need to evaluate it once we'll implement it.

File: kvbc/include/v4blockchain/v4_blockchain.h [hunk 2]
```diff
@@ -136,6 +125,105 @@ class KeyValueBlockchain {
   // Precondition3: `block_id_at_checkpoint` <= getLastReachableBlockId()
   void trimBlocksFromSnapshot(BlockId block_id_at_checkpoint);
 
+  //////////////////Recovery infra///////////////////////////
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/v4_blockchain.h:128]: May be later we should create a separate class and move the recovery infra outside of this class.

File: kvbc/include/v4blockchain/detail/latest_keys.h [hunk 3]
```diff
@@ -47,6 +47,8 @@ class LatestKeys {
   // E.L need to be used with compaction filter
   static constexpr Flags STALE_ON_UPDATE = {0x1};
   static constexpr size_t FLAGS_SIZE = STALE_ON_UPDATE.size();
+  static constexpr size_t TIME_STAMP_SIZE = sizeof(std::uint64_t);
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/detail/latest_keys.h]: Should we rename it to VERSION_SIZE, since the notion of the timestamp as the version is no longer required.

File: kvbc/src/v4blockchain/detail/latest_keys.cpp [hunk 4]
```diff
@@ -31,15 +30,12 @@ LatestKeys::LatestKeys(const std::shared_ptr<concord::storage::rocksdb::NativeCl
                        const std::optional<std::map<std::string, categorization::CATEGORY_TYPE>>& categories,
                        std::function<BlockId()>&& f)
     : native_client_{native_client}, category_mapping_(native_client, categories), comp_filter_(std::move(f)) {
-  if (native_client_->createColumnFamilyIfNotExisting(v4blockchain::detail::LATEST_KEYS_CF,
-                                                      concord::storage::rocksdb::getLexicographic64TsComparator(),
-                                                      getCompFilter())) {
+  if (native_client_->createColumnFamilyIfNotExisting(v4blockchain::detail::LATEST_KEYS_CF, getCompFilter())) {
     LOG_INFO(V4_BLOCK_LOG,
```
Comment (arc-vmware) [inline kvbc/src/v4blockchain/detail/latest_keys.cpp:34]: At some point later we should make it LOG_DEBUG

File: kvbc/src/v4blockchain/detail/latest_keys.cpp [hunk 5]
```diff
@@ -31,15 +30,12 @@ LatestKeys::LatestKeys(const std::shared_ptr<concord::storage::rocksdb::NativeCl
...
              "Created [" << v4blockchain::detail::LATEST_KEYS_CF << "] column family for the latest keys");
+    native_client_->put(v4blockchain::detail::LATEST_KEYS_CF, kvbc::keyTypes::v4_cf_flush, kvbc::V4Version());
```
Comment (arc-vmware) [inline kvbc/src/v4blockchain/detail/latest_keys.cpp]: Should this be kept in immutable CF, since this key need not be changed later?

File: kvbc/src/v4blockchain/detail/latest_keys.cpp [hunk 6]
```diff
@@ -74,15 +70,15 @@ void LatestKeys::handleCategoryUpdates(const std::string& block_version,
                             << concordUtils::bufferToHex(k.data(), k.size()) << " key size " << k.size()
                             << " value size " << v.size() << " raw key " << k);
     write_batch.put(
-        v4blockchain::detail::LATEST_KEYS_CF, getSliceArray(prefix, k, block_version), getSliceArray(v, sl_flags));
+        v4blockchain::detail::LATEST_KEYS_CF, getSliceArray(prefix, k), getSliceArray(v, sl_flags, block_version));
```
Comment (arc-vmware) [inline kvbc/src/v4blockchain/detail/latest_keys.cpp:72]: This is nice to keep the version with the value.

File: kvbc/src/v4blockchain/detail/latest_keys.cpp [hunk 7]
```diff
@@ -157,116 +153,124 @@ Iterate over updates, for each key:
...
     get_key.append(k);
-    // check if key exists for the previous version
-    auto opt_val = native_client_->get(cFamily, get_key, prev_block_version, &out_ts);
-    // if no previous version, delete it.
-    if (!opt_val) {
-      write_batch.del(cFamily, getSliceArray(prefix, k, block_version));
-      continue;
+    keys.push_back(get_key);
+  }
+  native_client_->multiGet(cFamily, keys, values, statuses, ro);
```
Comment (arc-vmware) [inline kvbc/src/v4blockchain/detail/latest_keys.cpp:208]: usage of multi get is nice here.

File: kvbc/src/v4blockchain/detail/latest_keys.cpp [hunk 8]
```diff
@@ -157,116 +153,124 @@ Iterate over updates, for each key:
...
-
-void LatestKeys::trimHistoryUntil(BlockId block_id) {
-  auto block_key = v4blockchain::detail::Blockchain::generateKey(block_id);
-  auto& raw_db = native_client_->rawDB();
-  auto status = raw_db.IncreaseFullHistoryTsLow(
-      native_client_->columnFamilyHandle(v4blockchain::detail::LATEST_KEYS_CF), block_key);
-  if (!status.ok()) {
-    LOG_ERROR(V4_BLOCK_LOG, "Failed trimming history due to " << status.ToString());
-    throw std::runtime_error(status.ToString());
+  native_client_->multiGet(v4blockchain::detail::LATEST_KEYS_CF, keys, values, statuses, ro);
```
Comment (arc-vmware) [inline kvbc/src/v4blockchain/detail/latest_keys.cpp:253]: usage of multi get is nice here.

File: kvbc/src/kvbc_adapter/replica_adapter.cpp [hunk 9]
```diff
@@ -59,7 +60,9 @@ ReplicaBlockchain::ReplicaBlockchain(
   if (aux_types.has_value()) {
     metrics_comp_.SetAggregator(aux_types->aggregator_);
   }
-  if (bftEngine::ReplicaConfig::instance().kvBlockchainVersion == BLOCKCHAIN_VERSION::CATEGORIZED_BLOCKCHAIN) {
+  auto blockchain_version = bftEngine::ReplicaConfig::instance().kvBlockchainVersion;
+  LOG_INFO(CAT_BLOCK_LOG, "Configured " << blockchain_version);
+  if (blockchain_version == BLOCKCHAIN_VERSION::CATEGORIZED_BLOCKCHAIN) {
```
Comment (arc-vmware) [inline kvbc/src/kvbc_adapter/replica_adapter.cpp:65]: change the else if aswell to
else if (blockchain_version == BLOCKCHAIN_VERSION::V4_BLOCKCHAIN)

File: kvbc/src/merkle_tree_storage_factory.cpp [hunk 10]
```diff
@@ -40,8 +40,8 @@ std::shared_ptr<rocksdb::Statistics> completeRocksDBConfiguration(
 
   // Use the same block cache and table options for all column familes for now.
   for (auto& d : cf_descs) {
-    auto* cf_table_options = reinterpret_cast<::rocksdb::BlockBasedTableOptions*>(
-        d.options.table_factory->GetOptions<::rocksdb::BlockBasedTableOptions>());
+    auto* cf_table_options =
+        reinterpret_cast<::rocksdb::BlockBasedTableOptions*>(d.options.table_factory->GetOptions());
```
Comment (arc-vmware) [inline kvbc/src/merkle_tree_storage_factory.cpp:44]: is this change needed?

**Review by arc-vmware — COMMENTED** [review]
Thanks @cloudnoize I complete the review of all the files. Now the recovery usecase is remaining. I will try to complete that by tonight.

File: kvbc/include/v4blockchain/detail/categories.h [hunk 11]
```diff
@@ -50,6 +50,15 @@ class Categories {
 
   std::map<std::string, concord::kvbc::categorization::CATEGORY_TYPE> getCategories() const { return category_types_; }
 
+  std::string getCategoryFromPrefix(const std::string& p) const {
+    for (const auto& [k, v] : category_to_prefix_) {
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/detail/categories.h:54]: nit: the category and prefix are 1:1 mapping. So it would be good to use a prefix_to_category_ map, since categories are finite and the space will be O(n) if we use 2 maps.
Another Idea is, as categories are unique and prefixes are also unique. And the category_to_prefix_ is a string to string map. So we can make it cat2prefix_prefix2cat_ map and we will insert cat->prefix and prefix->cat in the same map.

Comment (arc-vmware) [reply kvbc/include/v4blockchain/detail/categories.h:54]: nit cont: We can also use boost::bimap

File: kvbc/include/v4blockchain/detail/latest_keys.h [hunk 12]
```diff
@@ -66,47 +68,43 @@ class LatestKeys {
...
+                            concord::storage::rocksdb::NativeWriteBatch& write_batch,
+                            ::rocksdb::Snapshot* snpsht);
 
   const std::string& getCategoryPrefix(const std::string& category) const {
     return category_mapping_.categoryPrefix(category);
   }
-  // Mark a version that is safe to delete on compaction.
-  void trimHistoryUntil(BlockId block_id);
 
   static ::rocksdb::Slice getFlagsSlice(const ::rocksdb::Slice& val) {
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/detail/latest_keys.h:105]: nit: a convenience function called getVersionSlice might also be helpful.

File: kvbc/include/v4blockchain/v4_blockchain.h [hunk 13]
```diff
@@ -136,6 +125,105 @@ class KeyValueBlockchain {
...
+
+  The complexity is to revert the latest column family to the state prior the block addition, as
+  a key can get updated but its previous version is not known.
+
+  On BFT recovery, we detect that a BFT sequnce number has changed,
+  and we can take a rocksdb snapshot of its blockchain state in order to use for recovery if the current sn execution
+  fails.
+
+  the deletion process itself is a two phase operation:
+  1 - open the blockchain is readonly mode
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/v4_blockchain.h]: nit: open the blockchain (is -> in) readonly mode

File: kvbc/include/v4blockchain/v4_blockchain.h [hunk 14]
```diff
@@ -136,6 +125,105 @@ class KeyValueBlockchain {
...
+  were added duting a checkpoint creation.
+
+  The complexity is to revert the latest column family to the state prior the block addition, as
+  a key can get updated but its previous version is not known.
+
+  On BFT recovery, we detect that a BFT sequnce number has changed,
+  and we can take a rocksdb snapshot of its blockchain state in order to use for recovery if the current sn execution
+  fails.
+
+  the deletion process itself is a two phase operation:
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/v4_blockchain.h]: Should this line refer to Recovery Process or Deletion Process?

File: kvbc/include/v4blockchain/v4_blockchain.h [hunk 15]
```diff
@@ -136,6 +125,105 @@ class KeyValueBlockchain {
...
+  1 - open the blockchain is readonly mode
+  2 - get the updates needed for reverting the last block, using the snap shot of either BFT or checkpoint.
+  3 - save the serialized updates in a different recovery DB.
+  4- start the blockchain in write mode and apply the updates.
+
+  */
+
+  // take snapshot and releease the prev if BFT sn has changed.
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/v4_blockchain.h]: release

File: kvbc/include/v4blockchain/v4_blockchain.h [hunk 16]
```diff
@@ -136,6 +125,105 @@ class KeyValueBlockchain {
...
+  uint64_t onNewBFTSequenceNumber(const categorization::Updates &updates);
+  // take snapshot and releease on start and end of checkpoint.
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/v4_blockchain.h]: release

File: kvbc/include/v4blockchain/v4_blockchain.h [hunk 17]
```diff
@@ -136,6 +125,105 @@ class KeyValueBlockchain {
...
+      V4snapshot(::rocksdb::SequenceNumber sn) : sn_(sn) {}
+      virtual ::rocksdb::SequenceNumber GetSequenceNumber() const override { return sn_; }
+      virtual ~V4snapshot(){};
+      ::rocksdb::SequenceNumber sn_;
+    };
+
+    RecoverySnapshot(::rocksdb::DB *db) : db_(db) { sh_ = db_->GetSnapshot(); }
+    RecoverySnapshot(::rocksdb::DB *db, ::rocksdb::Snapshot *sh) : db_(db), sh_(sh) {}
+    /*
+    Get the value to store in the data base which is contactantion of the rocksdb sn,
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/v4_blockchain.h]: concatenation

File: kvbc/include/v4blockchain/v4_blockchain.h [hunk 18]
```diff
@@ -136,6 +125,105 @@ class KeyValueBlockchain {
   // Precondition3: `block_id_at_checkpoint` <= getLastReachableBlockId()
   void trimBlocksFromSnapshot(BlockId block_id_at_checkpoint);
 
+  //////////////////Recovery infra///////////////////////////
+  /*
+  Recovery is needed when we crash while executing a BFT sequence number's requests or when trimming the blocks that
+  were added duting a checkpoint creation.
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/v4_blockchain.h]: after the crash and before recovering, many times operation team does a metadata cleanup. This will move the sn of BFT to 0. Will this recovery work in those scenarios.

Comment (arc-vmware) [inline kvbc/include/v4blockchain/v4_blockchain.h]: duting -> during

File: kvbc/include/v4blockchain/v4_blockchain.h [hunk 19]
```diff
@@ -136,6 +125,105 @@ class KeyValueBlockchain {
...
+
+  The complexity is to revert the latest column family to the state prior the block addition, as
+  a key can get updated but its previous version is not known.
+
+  On BFT recovery, we detect that a BFT sequnce number has changed,
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/v4_blockchain.h]: sequnce -> sequence

File: kvbc/include/v4blockchain/v4_blockchain.h [hunk 20]
```diff
@@ -136,6 +125,105 @@ class KeyValueBlockchain {
...
+  // path - the path of the blockchain db.
+  struct BlockchainRecovery {
+    BlockchainRecovery(const std::string &path, bool is_chkpnt);
+    static std::shared_ptr<storage::rocksdb::NativeClient> getRecoveryDB(const std::string &path, bool is_chkpnt);
+    static void removeRecoveryDB(const std::string &path, bool is_chkpnt);
+    static std::string getPath(const std::string &path, bool is_chkpnt);
+  };
+
+  /*
+    this class is used on recovery, it's instantiated from the persisted RocksDB sequnce number,
```
Comment (arc-vmware) [inline kvbc/include/v4blockchain/v4_blockchain.h:162]: sequnce -> sequence

**Review by arc-vmware — COMMENTED** [review]
I will continue to review, couldn't complete the same by tonight.

File: kvbc/include/v4blockchain/v4_blockchain.h (hunk 18 above)
Comment (arc-vmware) [reply kvbc/include/v4blockchain/v4_blockchain.h]: Another que is, if the recovery process fails to create the recovery DB once, will it be possible to recover by restarting the replica. Currently, if we compare the recovery with categorized BC we will see that we can revert to a stable position even if there are crashes at the startup. So do we loose this property with this form of startup.

File: kvbc/include/v4blockchain/detail/latest_keys.h (hunk 3 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/detail/latest_keys.h]: ok

File: kvbc/include/v4blockchain/v4_blockchain.h (hunk 2 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/v4_blockchain.h:128]: agree

File: kvbc/src/v4blockchain/detail/latest_keys.cpp (hunk 7 above)
Comment (cloudnoize) [reply kvbc/src/v4blockchain/detail/latest_keys.cpp:208]: and it was the only way to use the snapshot when opening the database in read-only mode

File: kvbc/src/v4blockchain/detail/latest_keys.cpp (hunk 5 above)
Comment (cloudnoize) [reply kvbc/src/v4blockchain/detail/latest_keys.cpp]: actually, it's not needed anymore, thanks for spotting

File: kvbc/src/merkle_tree_storage_factory.cpp (hunk 10 above)
Comment (cloudnoize) [reply kvbc/src/merkle_tree_storage_factory.cpp:44]: it's a revert of a change needed for upgrading RocksDB

File: kvbc/include/v4blockchain/detail/categories.h (hunk 11 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/detail/categories.h:54]: yep, but I assume that the categories count is very limited and won't increase

File: kvbc/include/v4blockchain/v4_blockchain.h (hunk 13 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/v4_blockchain.h]: yep, I wrote those comments while was in the "zone" need to go over them and fix the grammar and typos

File: kvbc/include/v4blockchain/v4_blockchain.h (hunk 14 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/v4_blockchain.h]: I tried to describe how the actual delete works

File: kvbc/include/v4blockchain/v4_blockchain.h (hunk 15 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/v4_blockchain.h]: yep, I wrote those comments while was in the "zone" need to go over them and fix the grammar and typos

File: kvbc/include/v4blockchain/v4_blockchain.h (hunk 16 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/v4_blockchain.h]: yep, I wrote those comments while was in the "zone" need to go over them and fix the grammar and typos

File: kvbc/include/v4blockchain/v4_blockchain.h (hunk 17 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/v4_blockchain.h]: yep, I wrote those comments while was in the "zone" need to go over them and fix the grammar and typos

File: kvbc/include/v4blockchain/v4_blockchain.h (hunk 18 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/v4_blockchain.h]: if the metadata is cleared then the replica is considered to be in sync by the replica state sync and no blocks will be deleted.

the flow is resilient to crashes, let's go over it:
1- the recovery of requests and checkpoint will not use the same DB , the first will have path of “$rocksdbpath/recover_execution” and the other “$rocksdbpath/recover_chekpoint”
2 - When recovery is needed open an instance of the recover database.
3 - fill it with recovery data, do not override existing entries.
4 - now we open the blockchain in write mode
5 - delete blocks from the blockchain with recovery data
6 - on finish - delete the recovery instance
crashes:
if we crash in 3 - it will do it again, on next restart
if we crash in 4 - the write open is destructive, but the data is saved and will not be overwritten by 3.
if we crash in 5 - the data is still saved and we’ll not overwrite in 3 and it will continue from the last block id before the crash at 5.
if the deletion is successful we must call 6 and it guarantees that no garbage or stale data is left or used.

Comment (cloudnoize) [reply kvbc/include/v4blockchain/v4_blockchain.h]: yep, I wrote those comments while was in the "zone" need to go over them and fix the grammar and typos

File: kvbc/include/v4blockchain/v4_blockchain.h (hunk 19 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/v4_blockchain.h]: yep, I wrote those comments while was in the "zone" need to go over them and fix the grammar and typos

File: kvbc/include/v4blockchain/v4_blockchain.h (hunk 20 above)
Comment (cloudnoize) [reply kvbc/include/v4blockchain/v4_blockchain.h:162]: yep, I wrote those comments while was in the "zone" need to go over them and fix the grammar and typos
